tmt/steps/execute/internal.py

```diff
@@ -219,48 +378,111 @@ def check(self, test):
 
     def _setup_reboot_script(self, guest):
         """ Set up reboot script on the guest. """
-        def setup_file(filename, content):
-            file_path = os.path.join(self.workdir, filename)
-            with open(file_path, 'w') as template:
-                template.write(content)
-            # Make it executable, keep perms when pushing
-            perms = os.stat(file_path)
-            os.chmod(file_path, perms.st_mode | stat.S_IEXEC)
-            return file_path
-
-        template = setup_file(REBOOT_TEMPLATE_NAME, REBOOT_SCRIPT)
-        setup = setup_file(REBOOT_SETUP_NAME, REBOOT_SETUP_SCRIPT)
-        teardown = setup_file(REBOOT_TEARDOWN_NAME, REBOOT_TEARDOWN_SCRIPT)
+
+        template = self._setup_tmt_restraint_file(REBOOT_TEMPLATE_NAME,
+                                                  REBOOT_SCRIPT,
+                                                  REBOOT_REQUEST_FILENAME)
+        setup = self._setup_tmt_restraint_file(REBOOT_SETUP_NAME,
+                                               TMT_RESTRAINT_SETUP_SCRIPT,
+                                               REBOOT_SCRIPT_PATHS)
+        teardown = \
+            self._setup_tmt_restraint_file(REBOOT_TEARDOWN_NAME,
+                                           TMT_RESTRAINT_TEARDOWN_SCRIPT,
+                                           REBOOT_SCRIPT_PATHS)
+        guest.push(self.workdir, options=DEFAULT_RSYNC_OPTIONS + ["-p"])
+        return template, setup, teardown
+
+    def _setup_rstrnt_report_result_script(self, guest):
+        """ Set up rstrnt-report-result script on the guest. """
+        template = \
+            self._setup_tmt_restraint_file(REPORT_RESULT_TEMPLATE_NAME,
+                                           RSTRNT_REPORT_RESULT_SCRIPT)
+        setup = \
```

Comment:
So this got me thinking, wouldn't it be possible to use a single setup and teardown script for logic inside both reboot and report-result? It could take the files to replace from stdin or options.